File: 2026/07/08/bowling-score/haskell/src/Bowling.hs
module Bowling ( score )
    where

type Roll = Int
type Bonus = Int
data Frame = Frame Roll Roll 
           | HalfFrame Roll
           | Spare Roll Bonus
           | Strike Bonus
    deriving (Eq, Show)

extra :: Int -> [Roll] -> Int
extra n = sum . take n

frames :: Int -> [Roll] -> [Frame]
frames 10 _ = []
frames _ [] = []
frames _ [roll] = [HalfFrame roll]
frames n (roll:rolls) | roll == 10
  = Strike (2 `extra` rolls) : frames (succ n) rolls

frames n (rollA:rollB:rolls) | rollA + rollB == 10
  = Spare rollA (1 `extra` rolls) : frames (succ n) rolls

frames n (rollA:rollB:rolls)
  = Frame rollA rollB : frames (succ n) rolls

frameScore :: Frame -> Int
frameScore (HalfFrame roll) = roll
frameScore (Frame rollA rollB) = rollA + rollB
frameScore (Spare _ bonus) = 10 + bonus
frameScore (Strike bonus) = 10 + bonus

score :: [Roll] -> Int
score = sum . map frameScore . frames 0

File: 2026/07/08/bowling-score/c/src/bowling.c
// bowling.c
#include "bowling.h"
#include <stdio.h>

int score(int *rolls, int nb_rolls) {
    int score = 0;
    int bonus = 0;
    int next_bonus = 0;
    int last_roll = 0;
    int half_frame = 0;
    for (int index = 0; index < nb_rolls; index++) {
        int roll = rolls[index];
        score += roll * bonus;
        bonus = next_bonus;
        next_bonus = 0;
        if (half_frame < 20) {
            if (half_frame % 2 == 0) {
                if (roll == 10) {
                    bonus++;
                    next_bonus++;
                    half_frame++;
                } else {
                    last_roll = roll;
                }
            }
            else {
                if (last_roll + roll == 10) {
                    bonus = 1;
                }
            }
            score += roll;
            half_frame++;
        }
    }
    return score;
}
